File: packages/python/python/sombra/typed.py
from typing import Any, Generic, TypeVar, overload, TYPE_CHECKING

if TYPE_CHECKING:
    from typing_extensions import TypedDict
    from .sombra import (
        SombraDB as NativeSombraDB,
        SombraTransaction as NativeSombraTransaction,
        SombraNode,
        SombraEdge,
        BfsResult,
        QueryBuilder as NativeQueryBuilder,
    )
else:
    try:
        from typing_extensions import TypedDict
    except ImportError:
        from typing import TypedDict
    
    from .sombra import (
        SombraDB as NativeSombraDB,
        SombraTransaction as NativeSombraTransaction,
        SombraNode,
        SombraEdge,
        BfsResult,
        QueryBuilder as NativeQueryBuilder,
    )

PropertyValue = bool | int | float | str | bytes

SchemaT = TypeVar('SchemaT', bound='GraphSchema')


class GraphSchema(TypedDict):
    nodes: dict[str, dict[str, PropertyValue]]
    edges: dict[str, Any]


class TypedQueryBuilder(Generic[SchemaT]):
    def __init__(self, db: 'SombraDB[SchemaT]', builder: 'NativeQueryBuilder'):
        self._db = db
        self._builder = builder

    def start_from(self, node_ids: list[int]) -> 'TypedQueryBuilder[SchemaT]':
        self._builder.start_from(node_ids)
        return self

    def start_from_label(self, label: str) -> 'TypedQueryBuilder[SchemaT]':
        self._builder.start_from_label(label)
        return self

    def start_from_property(
        self,
        label: str,
        key: str,
        value: PropertyValue
    ) -> 'TypedQueryBuilder[SchemaT]':
        self._builder.start_from_property(label, key, value)
        return self

    def traverse(
        self,
        edge_types: list[str],
        direction: str,
        depth: int
    ) -> 'TypedQueryBuilder[SchemaT]':
        self._builder.traverse(edge_types, direction, depth)
        return self

    def limit(self, n: int) -> 'TypedQueryBuilder[SchemaT]':
        self._builder.limit(n)
        return self

    def execute(self) -> Any:
        return self._builder.execute()


class SombraTransaction(Generic[SchemaT]):
    def __init__(self, tx: 'NativeSombraTransaction'):
        self._tx = tx

    def id(self) -> int:
        return self._tx.id()

    @overload
    def add_node(self, labels: str, properties: dict[str, PropertyValue]) -> int: ...
    
    @overload
    def add_node(self, labels: list[str], properties: dict[str, PropertyValue]) -> int: ...

    def add_node(
        self,
        labels: str | list[str],
        properties: dict[str, PropertyValue]
    ) -> int:
        if isinstance(labels, str):
            labels = [labels]
        return self._tx.add_node(labels, properties)

    def add_edge(
        self,
        source_node_id: int,
        target_node_id: int,
        edge_type: str,
        properties: dict[str, PropertyValue] | None = None
    ) -> int:
        return self._tx.add_edge(
            source_node_id,
            target_node_id,
            edge_type,
            properties or {}
        )

    def get_node(self, node_id: int) -> 'SombraNode':
        return self._tx.get_node(node_id)

    def get_edge(self, edge_id: int) -> 'SombraEdge':
        return self._tx.get_edge(edge_id)

    def get_neighbors(self, node_id: int) -> list[int]:
        return self._tx.get_neighbors(node_id)

    def get_outgoing_edges(self, node_id: int) -> list[int]:
        return self._tx.get_outgoing_edges(node_id)

    def get_incoming_edges(self, node_id: int) -> list[int]:
        return self._tx.get_incoming_edges(node_id)

    def delete_node(self, node_id: int) -> None:
        self._tx.delete_node(node_id)

    def delete_edge(self, edge_id: int) -> None:
        self._tx.delete_edge(edge_id)

    def set_node_property(
        self,
        node_id: int,
        key: str,
        value: PropertyValue
    ) -> None:
        self._tx.set_node_property(node_id, key, value)

    def remove_node_property(self, node_id: int, key: str) -> None:
        self._tx.remove_node_property(node_id, key)

    def commit(self) -> None:
        self._tx.commit()

    def rollback(self) -> None:
        self._tx.rollback()

    def get_incoming_neighbors(self, node_id: int) -> list[int]:
        return self._tx.get_incoming_neighbors(node_id)

    def get_neighbors_two_hops(self, node_id: int) -> list[int]:
        return self._tx.get_neighbors_two_hops(node_id)

    def get_neighbors_three_hops(self, node_id: int) -> list[int]:
        return self._tx.get_neighbors_three_hops(node_id)

    def bfs_traversal(self, start_node_id: int, max_depth: int) -> list['BfsResult']:
        return self._tx.bfs_traversal(start_node_id, max_depth)

    def get_nodes_by_label(self, label: str) -> list[int]:
        return self._tx.get_nodes_by_label(label)

    def get_nodes_in_range(self, start: int, end: int) -> list[int]:
        return self._tx.get_nodes_in_range(start, end)

    def get_nodes_from(self, start: int) -> list[int]:
        return self._tx.get_nodes_from(start)

    def get_nodes_to(self, end: int) -> list[int]:
        return self._tx.get_nodes_to(end)

    def get_first_node(self) -> int | None:
        return self._tx.get_first_node()

    def get_last_node(self) -> int | None:
        return self._tx.get_last_node()

    def get_first_n_nodes(self, n: int) -> list[int]:
        return self._tx.get_first_n_nodes(n)

    def get_last_n_nodes(self, n: int) -> list[int]:
        return self._tx.get_last_n_nodes(n)

    def get_all_node_ids_ordered(self) -> list[int]:
        return self._tx.get_all_node_ids_ordered()

    def count_outgoing_edges(self, node_id: int) -> int:
        return self._tx.count_outgoing_edges(node_id)

    def count_incoming_edges(self, node_id: int) -> int:
        return self._tx.count_incoming_edges(node_id)


class SombraDB(Generic[SchemaT]):
    def __init__(self, path: str):
        self._db = NativeSombraDB(path)

    def begin_transaction(self) -> SombraTransaction[SchemaT]:
        return SombraTransaction(self._db.begin_transaction())

    @overload
    def add_node(self, labels: str, properties: dict[str, PropertyValue]) -> int: ...
    
    @overload
    def add_node(self, labels: list[str], properties: dict[str, PropertyValue]) -> int: ...

    def add_node(
        self,
        labels: str | list[str],
        properties: dict[str, PropertyValue]
    ) -> int:
        if isinstance(labels, str):
            labels = [labels]
        return self._db.add_node(labels, properties)

    def add_edge(
        self,
        source_node_id: int,
        target_node_id: int,
        edge_type: str,
        properties: dict[str, PropertyValue] | None = None
    ) -> int:
        return self._db.add_edge(
            source_node_id,
            target_node_id,
            edge_type,
            properties or {}
        )

    def get_node(self, node_id: int) -> 'SombraNode':
        return self._db.get_node(node_id)

    def get_edge(self, edge_id: int) -> 'SombraEdge':
        return self._db.get_edge(edge_id)

    def get_neighbors(self, node_id: int) -> list[int]:
        return self._db.get_neighbors(node_id)

    def get_outgoing_edges(self, node_id: int) -> list[int]:
        return self._db.get_outgoing_edges(node_id)

    def get_incoming_edges(self, node_id: int) -> list[int]:
        return self._db.get_incoming_edges(node_id)

    def delete_node(self, node_id: int) -> None:
        self._db.delete_node(node_id)

    def delete_edge(self, edge_id: int) -> None:
        self._db.delete_edge(edge_id)

    def set_node_property(
        self,
        node_id: int,
        key: str,
        value: PropertyValue
    ) -> None:
        self._db.set_node_property(node_id, key, value)

    def remove_node_property(self, node_id: int, key: str) -> None:
        self._db.remove_node_property(node_id, key)

    def flush(self) -> None:
        self._db.flush()

    def checkpoint(self) -> None:
        self._db.checkpoint()

    def get_incoming_neighbors(self, node_id: int) -> list[int]:
        return self._db.get_incoming_neighbors(node_id)

    def get_neighbors_two_hops(self, node_id: int) -> list[int]:
        return self._db.get_neighbors_two_hops(node_id)

    def get_neighbors_three_hops(self, node_id: int) -> list[int]:
        return self._db.get_neighbors_three_hops(node_id)

    def bfs_traversal(self, start_node_id: int, max_depth: int) -> list['BfsResult']:
        return self._db.bfs_traversal(start_node_id, max_depth)

    def get_nodes_by_label(self, label: str) -> list[int]:
        return self._db.get_nodes_by_label(label)

    def get_nodes_in_range(self, start: int, end: int) -> list[int]:
        return self._db.get_nodes_in_range(start, end)

    def get_nodes_from(self, start: int) -> list[int]:
        return self._db.get_nodes_from(start)

    def get_nodes_to(self, end: int) -> list[int]:
        return self._db.get_nodes_to(end)

    def get_first_node(self) -> int | None:
        return self._db.get_first_node()

    def get_last_node(self) -> int | None:
        return self._db.get_last_node()

    def get_first_n_nodes(self, n: int) -> list[int]:
        return self._db.get_first_n_nodes(n)

    def get_last_n_nodes(self, n: int) -> list[int]:
        return self._db.get_last_n_nodes(n)

    def get_all_node_ids_ordered(self) -> list[int]:
        return self._db.get_all_node_ids_ordered()

    def count_outgoing_edges(self, node_id: int) -> int:
        return self._db.count_outgoing_edges(node_id)

    def count_incoming_edges(self, node_id: int) -> int:
        return self._db.count_incoming_edges(node_id)

    def query(self) -> TypedQueryBuilder[SchemaT]:
        return TypedQueryBuilder(self, self._db.query())


__all__ = [
    'SombraDB',
    'SombraTransaction',
    'TypedQueryBuilder',
    'GraphSchema',
    'PropertyValue',
]

File: packages/nodejs/typed.js
const { SombraDB: NativeSombraDB } = require("./index.js");

/**
 * Converts a JavaScript primitive value to SombraDB property format.
 * @param {string|number|boolean} value - The value to convert
 * @returns {{type: 'string'|'int'|'float'|'bool', value: any}} SombraDB property object
 */
function convertToSombraProperty(value) {
	if (typeof value === "string") {
		return { type: "string", value };
	} else if (typeof value === "number") {
		if (Number.isInteger(value)) {
			return { type: "int", value };
		} else {
			return { type: "float", value };
		}
	} else if (typeof value === "boolean") {
		return { type: "bool", value };
	}
	throw new Error(`Unsupported property type: ${typeof value}`);
}

/**
 * Converts a SombraDB property object to a JavaScript primitive value.
 * @param {{type: string, value: any}} sombraValue - The SombraDB property to convert
 * @returns {string|number|boolean} The primitive value
 */
function convertFromSombraProperty(sombraValue) {
	if (!sombraValue || typeof sombraValue !== "object") {
		return sombraValue;
	}
	return sombraValue.value;
}

/**
 * Converts an object of JavaScript primitive values to SombraDB property format.
 * @param {Object.<string, string|number|boolean>} properties - Properties to convert
 * @returns {Object.<string, {type: string, value: any}>} SombraDB properties
 */
function convertPropertiesToSombra(properties) {
	const result = {};
	for (const [key, value] of Object.entries(properties)) {
		result[key] = convertToSombraProperty(value);
	}
	return result;
}

/**
 * Converts SombraDB properties to JavaScript primitive values.
 * @param {Object.<string, {type: string, value: any}>} sombraProperties - SombraDB properties
 * @returns {Object.<string, string|number|boolean>} JavaScript properties
 */
function convertPropertiesFromSombra(sombraProperties) {
	const result = {};
	for (const [key, value] of Object.entries(sombraProperties)) {
		result[key] = convertFromSombraProperty(value);
	}
	return result;
}

function isTypedCall(labelOrLabels, properties) {
	// Check if it's a typed call (not raw API with {type, value} properties)
	if (!properties || typeof properties !== "object") {
		return false;
	}

	// Check if properties are in raw format
	const firstValue = Object.values(properties)[0];
	if (isTypedPropertyValue(firstValue)) {
		return false;
	}

	// Accept string or array of strings for labels
	if (typeof labelOrLabels === "string") {
		return true;
	}

	if (Array.isArray(labelOrLabels) && labelOrLabels.length > 0) {
		return labelOrLabels.every((label) => typeof label === "string");
	}

	return false;
}

function isTypedPropertyValue(value) {
	return (
		value && typeof value === "object" && "type" in value && "value" in value
	);
}

class TypedQueryBuilderImpl {
	constructor(db, builder) {
		this._db = db;
		this._builder = builder;
	}

	startFrom(nodeIds) {
		this._builder.startFrom(nodeIds);
		return this;
	}

	startFromLabel(label) {
		this._builder.startFromLabel(label);
		return this;
	}

	startFromProperty(label, key, value) {
		this._builder.startFromProperty(label, key, String(value));
		return this;
	}

	traverse(edgeTypes, direction, depth) {
		this._builder.traverse(edgeTypes, direction, depth);
		return this;
	}

	limit(n) {
		this._builder.limit(n);
		return this;
	}

	getIds() {
		return this._builder.getIds();
	}

	getNodes() {
		const nodes = this._builder.getNodes();
		return nodes.map((node) => ({
			id: node.id,
			labels: node.labels,
			properties: convertPropertiesFromSombra(node.properties),
		}));
	}

	execute() {
		return this._builder.getIds();
	}
}

/**
 * Unified SombraDB API class that works with or without TypeScript generics.
 *
 * @class
 * @example
 * // Type-safe usage with TypeScript
 * const db = new SombraDB<MySchema>('./db');
 * db.addNode('Person', { name: 'Alice', age: 30 });
 *
 * @example
 * // Raw usage (backwards compatible)
 * const db = new SombraDB('./db');
 * db.addNode(['Person'], { name: { type: 'string', value: 'Alice' } });
 */
class SombraDB {
	/**
	 * Creates a new SombraDB instance.
	 * @param {string} path - Path to the database file
	 */
	constructor(path) {
		this._db = new NativeSombraDB(path);
	}

	/**
	 * Get access to the underlying native SombraDB instance.
	 * @returns {NativeSombraDB} The native database instance
	 */
	get db() {
		return this._db;
	}

	/**
	 * Begin a new transaction.
	 * @returns {SombraTransaction} The transaction object
	 */
	beginTransaction() {
		return this._db.beginTransaction();
	}

	/**
	 * Add a node to the graph.
	 * @param {string|string[]} labelOrLabels - Node label (typed) or array of labels (typed/raw)
	 * @param {Object} properties - Node properties (plain values for typed, {type, value} for raw)
	 * @returns {number} The node ID
	 */
	addNode(labelOrLabels, properties = {}) {
		if (isTypedCall(labelOrLabels, properties)) {
			const sombraProps = convertPropertiesToSombra(properties);
			const labels = Array.isArray(labelOrLabels)
				? labelOrLabels
				: [labelOrLabels];
			return this._db.addNode(labels, sombraProps);
		}
		return this._db.addNode(labelOrLabels, properties);
	}

	/**
	 * Add an edge between two nodes.
	 * @param {number} sourceNodeId - Source node ID
	 * @param {number} targetNodeId - Target node ID
	 * @param {string} edgeTypeOrLabel - Edge type name
	 * @param {Object} properties - Edge properties (plain values for typed, {type, value} for raw)
	 * @returns {number} The edge ID
	 */
	addEdge(sourceNodeId, targetNodeId, edgeTypeOrLabel, properties = {}) {
		if (properties && Object.keys(properties).length > 0) {
			const firstValue = Object.values(properties)[0];
			if (isTypedPropertyValue(firstValue)) {
				return this._db.addEdge(
					sourceNodeId,
					targetNodeId,
					edgeTypeOrLabel,
					properties,
				);
			}
			const sombraProps = convertPropertiesToSombra(properties);
			return this._db.addEdge(
				sourceNodeId,
				targetNodeId,
				edgeTypeOrLabel,
				sombraProps,
			);
		}
		return this._db.addEdge(
			sourceNodeId,
			targetNodeId,
			edgeTypeOrLabel,
			properties,
		);
	}

	/**
	 * Get a node by ID. Properties are always returned as plain JavaScript values.
	 * @param {number} nodeId - The node ID
	 * @returns {Object|null} The node object with plain property values, or null if not found
	 */
	getNode(nodeId) {
		try {
			const node = this._db.getNode(nodeId);
			if (!node) return null;

			return {
				id: node.id,
				labels: node.labels,
				properties: convertPropertiesFromSombra(node.properties),
			};
		} catch (e) {
			return null;
		}
	}

	/**
	 * Get all node IDs with a specific label.
	 * @param {string} label - The node label
	 * @returns {number[]} Array of node IDs
	 */
	getNodesByLabel(label) {
		return this._db.getNodesByLabel(label);
	}

	/**
	 * Find the first node matching a label and property value.
	 * @param {string} label - The node label
	 * @param {string} key - Property key
	 * @param {string|number|boolean} value - Property value to match
	 * @returns {number|undefined} The node ID, or undefined if not found
	 */
	findNodeByProperty(label, key, value) {
		const result = this._db
			.query()
			.startFromProperty(label, key, String(value))
			.execute();
		return result.nodeIds[0];
	}

	/**
	 * Find all nodes matching a label and property value.
	 * @param {string} label - The node label
	 * @param {string} key - Property key
	 * @param {string|number|boolean} value - Property value to match
	 * @returns {number[]} Array of node IDs
	 */
	findNodesByProperty(label, key, value) {
		const result = this._db
			.query()
			.startFromProperty(label, key, String(value))
			.execute();
		return result.nodeIds;
	}

	/**
	 * Get an edge by ID. Properties are always returned as plain JavaScript values.
	 * @param {number} edgeId - The edge ID
	 * @returns {Object|null} The edge object with plain property values, or null if not found
	 */
	getEdge(edgeId) {
		try {
			const edge = this._db.getEdge(edgeId);
			if (!edge) return null;

			return {
				id: edge.id,
				sourceNodeId: edge.sourceNodeId,
				targetNodeId: edge.targetNodeId,
				typeName: edge.typeName,
				properties: convertPropertiesFromSombra(edge.properties),
			};
		} catch (e) {
			return null;
		}
	}

	/**
	 * Get all outgoing edge IDs from a node.
	 * @param {number} nodeId - The node ID
	 * @returns {number[]} Array of edge IDs
	 */
	getOutgoingEdges(nodeId) {
		return this._db.getOutgoingEdges(nodeId);
	}

	/**
	 * Get all incoming edge IDs to a node.
	 * @param {number} nodeId - The node ID
	 * @returns {number[]} Array of edge IDs
	 */
	getIncomingEdges(nodeId) {
		return this._db.getIncomingEdges(nodeId);
	}

	/**
	 * Get all neighboring node IDs (outgoing connections).
	 * @param {number} nodeId - The node ID
	 * @returns {number[]} Array of node IDs
	 */
	getNeighbors(nodeId) {
		return this._db.getNeighbors(nodeId);
	}

	/**
	 * Get all incoming neighboring node IDs.
	 * @param {number} nodeId - The node ID
	 * @returns {number[]} Array of node IDs
	 */
	getIncomingNeighbors(nodeId) {
		return this._db.getIncomingNeighbors(nodeId);
	}

	/**
	 * Delete a node and all its connected edges.
	 * @param {number} nodeId - The node ID to delete
	 */
	deleteNode(nodeId) {
		return this._db.deleteNode(nodeId);
	}

	/**
	 * Delete an edge.
	 * @param {number} edgeId - The edge ID to delete
	 */
	deleteEdge(edgeId) {
		return this._db.deleteEdge(edgeId);
	}

	/**
	 * Set a property on a node.
	 * @param {number} nodeId - The node ID
	 * @param {string} key - Property key
	 * @param {string|number|boolean|Object} value - Property value (plain value for typed, {type, value} for raw)
	 */
	setNodeProperty(nodeId, key, value) {
		if (isTypedPropertyValue(value)) {
			return this._db.setNodeProperty(nodeId, key, value);
		}
		const sombraValue = convertToSombraProperty(value);
		return this._db.setNodeProperty(nodeId, key, sombraValue);
	}

	/**
	 * Remove a property from a node.
	 * @param {number} nodeId - The node ID
	 * @param {string} key - Property key to remove
	 */
	removeNodeProperty(nodeId, key) {
		return this._db.removeNodeProperty(nodeId, key);
	}

	/**
	 * Flush pending writes to disk.
	 */
	flush() {
		return this._db.flush();
	}

	/**
	 * Create a checkpoint of the current database state.
	 */
	checkpoint() {
		return this._db.checkpoint();
	}

	/**
	 * Perform a breadth-first search traversal.
	 * @param {number} startNodeId - Starting node ID
	 * @param {number} maxDepth - Maximum depth to traverse
	 * @returns {Array<{nodeId: number, depth: number}>} Array of nodes with their depths
	 */
	bfsTraversal(startNodeId, maxDepth) {
		return this._db.bfsTraversal(startNodeId, maxDepth);
	}

	/**
	 * Create a new query builder for complex graph queries.
	 * @returns {TypedQueryBuilderImpl} Query builder instance
	 */
	query() {
		const builder = this._db.query();
		return new TypedQueryBuilderImpl(this._db, builder);
	}

	/**
	 * Count nodes grouped by label.
	 * @returns {Object.<string, number>} Object mapping labels to counts
	 */
	countNodesByLabel() {
		return this._db.countNodesByLabel();
	}

	/**
	 * Count edges grouped by type.
	 * @returns {Object.<string, number>} Object mapping edge types to counts
	 */
	countEdgesByType() {
		return this._db.countEdgesByType();
	}

	/**
	 * Count nodes with a specific label.
	 * @param {string} label - The node label
	 * @returns {number} Count of nodes
	 */
	countNodesWithLabel(label) {
		return this._db.countNodesWithLabel(label);
	}

	/**
	 * Count edges with a specific type.
	 * @param {string} edgeType - The edge type
	 * @returns {number} Count of edges
	 */
	countEdgesWithType(edgeType) {
		return this._db.countEdgesWithType(edgeType);
	}

	/**
	 * Get all node IDs in order.
	 * @returns {number[]} Array of all node IDs
	 */
	getAllNodeIdsOrdered() {
		return this._db.getAllNodeIdsOrdered();
	}

	/**
	 * Get the first node ID in the database.
	 * @returns {number|null} First node ID or null if database is empty
	 */
	getFirstNode() {
		return this._db.getFirstNode();
	}

	/**
	 * Get the last node ID in the database.
	 * @returns {number|null} Last node ID or null if database is empty
	 */
	getLastNode() {
		return this._db.getLastNode();
	}

	/**
	 * Get all ancestor nodes by traversing edges backwards.
	 * @param {number} startNodeId - Starting node ID
	 * @param {string} edgeType - Edge type to traverse
	 * @param {number} [maxDepth] - Maximum depth (optional)
	 * @returns {number[]} Array of ancestor node IDs
	 */
	getAncestors(startNodeId, edgeType, maxDepth) {
		return this._db.getAncestors(startNodeId, edgeType, maxDepth);
	}

	/**
	 * Get all descendant nodes by traversing edges forwards.
	 * @param {number} startNodeId - Starting node ID
	 * @param {string} edgeType - Edge type to traverse
	 * @param {number} [maxDepth] - Maximum depth (optional)
	 * @returns {number[]} Array of descendant node IDs
	 */
	getDescendants(startNodeId, edgeType, maxDepth) {
		return this._db.getDescendants(startNodeId, edgeType, maxDepth);
	}

	/**
	 * Find the shortest path between two nodes.
	 * @param {number} start - Start node ID
	 * @param {number} end - End node ID
	 * @param {string[]} [edgeTypes] - Optional edge types to traverse
	 * @returns {number[]|null} Array of node IDs in path, or null if no path exists
	 */
	shortestPath(start, end, edgeTypes) {
		return this._db.shortestPath(start, end, edgeTypes);
	}

	/**
	 * Find all paths between two nodes within depth constraints.
	 * @param {number} start - Start node ID
	 * @param {number} end - End node ID
	 * @param {number} minDepth - Minimum path depth
	 * @param {number} maxDepth - Maximum path depth
	 * @param {string[]} [edgeTypes] - Optional edge types to traverse
	 * @returns {number[][]} Array of paths, where each path is an array of node IDs
	 */
	findPaths(start, end, minDepth, maxDepth, edgeTypes) {
		return this._db.findPaths(start, end, minDepth, maxDepth, edgeTypes);
	}

	/**
	 * Verify database integrity.
	 * @param {Object} options - Verification options
	 * @param {boolean} [options.checksumOnly] - Only verify checksums
	 * @param {boolean} [options.verifyIndexes] - Verify indexes (default: true)
	 * @param {boolean} [options.verifyAdjacency] - Verify adjacency (default: true)
	 * @param {number} [options.maxErrors] - Maximum errors to report (default: 16)
	 * @returns {Object} Integrity report
	 */
	verifyIntegrity(options) {
		return this._db.verifyIntegrity(options);
	}

	/**
	 * Get database header state.
	 * @returns {Object} Header state
	 */
	getHeader() {
		return this._db.getHeader();
	}

	/**
	 * Get database metrics.
	 * @returns {Object} Metrics
	 */
	getMetrics() {
		return this._db.getMetrics();
	}
}

/**
 * Get the default page size used by the database.
 * @returns {number} Page size in bytes
 */
function getDefaultPageSize() {
	const native = require("./index.js");
	return native.getDefaultPageSize();
}

module.exports = {
	SombraDB,
	getDefaultPageSize,
};
